src/storage/metrics.rs

```diff
@@ -134,6 +140,40 @@ pub fn tls_collect_read_flow(region_id: u64, statistics: &Statistics) {
     });
 }
 
+pub fn tls_collect_qps(
+    region_id: u64,
+    peer: &metapb::Peer,
+    start_key: &[u8],
+    end_key: &[u8],
+    reverse_scan: bool,
+) {
+    TLS_STORAGE_METRICS.with(|m| {
+        let mut m = m.borrow_mut();
+        let mut key_range = build_key_range(start_key, end_key, false);
+        if reverse_scan {
+            std::mem::swap(&mut key_range.start_key, &mut key_range.end_key)
+        }
+        m.local_qps_stats.add(region_id, peer, key_range);
+    });
+}
+
+pub fn tls_collect_qps_batch(
+    region_id: u64,
+    peer: &metapb::Peer,
+    mut key_ranges: Vec<KeyRange>,
+    reverse_scan: bool,
+) {
+    TLS_STORAGE_METRICS.with(|m| {
+        let mut m = m.borrow_mut();
+        if reverse_scan {
+            for range in key_ranges.iter_mut() {
+                std::mem::swap(&mut range.start_key, &mut range.end_key)
```

Comment:
use `build_key_range`?

Reply:
`reverse_scan` might be `true` only in https://github.com/tikv/tikv/blob/eacae77b9712812c1f62202d8cf4db44b2385b6a/src/storage/mod.rs#L997 and we directly get `range.clone` from there.

Reply:
Why need a clone or swap?   I think that use the `build_key_range` can reduce the memory copy on the hot path?